File: twopointers/0557-reverse-words-in-a-string-iii.go
package twopointers

import (
	"slices"
	"strings"
)

// https://leetcode.com/problems/reverse-words-in-a-string-iii/

// Approach: built-in functions
// Time: O(n)
// Space: O(n)
func reverseWords(s string) string {
	var words = strings.Split(s, " ")
	for idx := range words {
		word := []byte(words[idx])
		slices.Reverse(word)
		words[idx] = string(word)
	}
	return strings.Join(words, " ")
}

// // Approach: Two Pointers
// // Time: O(n)
// // Space: O(1)
// func reverseWords(s string) string {
// 	var b = []byte(s)
// 	var start, n = 0, len(s)
// 	for end := range s {
// 		if s[end] == ' ' || end == n-1 {
// 			var last = end
// 			if s[end] == ' ' {
// 				end--
// 			}
// 			for start < end {
// 				b[start], b[end] = b[end], b[start]
// 				start++
// 				end--
// 			}
// 			start = last + 1
// 		}
// 	}
// 	return string(b)
// }

File: twopointers/0557-reverse-words-in-a-string-iii.py
import unittest

# https://leetcode.com/problems/reverse-words-in-a-string-iii/

# python3 -m unittest twopointers/0557-reverse-words-in-a-string-iii.py


class Solution(unittest.TestCase):
    # # Approach: Two Pointers
    # # Time: O(n)
    # # Space: O(n)
    # def reverserWords(self, s: str) -> str:
    #     s = list(s) # [c for c in s]
    #     start = 0
    #     n = len(s)
    #     for end in range(n):
    #         if s[end] == ' ' or end == n-1:
    #             temp = end
    #             end -= int(s[end] == ' ')
    #             while start < end:
    #                 s[start], s[end] = s[end], s[start]
    #                 start += 1
    #                 end -= 1
    #             start = temp+1
    #     return "".join(s)

    # Approach: One-Liner
    # Time: O(n)
    # Space: O(n)
    def reverserWords(self, s: str) -> str:
        return " ".join([w[::-1] for w in s.split(" ")])

    def testReverserWords(self) -> None:
        for s, expected in [
            ("Let's take LeetCode contest", "s'teL ekat edoCteeL tsetnoc"),
            ("Mr Ding", "rM gniD"),
        ]:
            output = self.reverserWords(s)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")
